File: c++/algorithm/KMP_알고리즘.cpp
#include <iostream>
#include <random>
#include <vector>

#define TEST_CASES 10
#define HSIZE 1000000
#define MSIZE 10

using namespace std;

void initialize(string& s, string& m, default_random_engine& gen) {
  uniform_int_distribution<int> dis('a', 'c');

  for (int i = 0; i < HSIZE; i++) s += dis(gen);
  for (int i = 0; i < MSIZE; i++) m += dis(gen);
}

vector<int> get_partial_match(const string& m) {
  int M = m.size();
  vector<int> pi(M, 0);

  for (int being = 1; begin < M; begin++) {
    for (int i = 0; i < M; i++) {
      if (m[begin + i] != m[i]) break;
      pi[begin + i] = max(pi[begin + i], i + 1);
    }
  }
  return pi;
}

vector<int> solve(const string& h, const string& m) {
  int H = h.size(), M = m.size();
  vector<int> ret;
  vector<int> pi = get_partial_match(m);

  int begin = 0, matched = 0;

  while (begin <= H - M) {
    if (matched < M && h[begin + matched] == m[matched]) {
      ++matched;

      if (matched == M) ret.push_back(begin);
    } else {
      if (matched == 0)
        ++begin;
      else {
        begin += matched - pi[matched - 1];
        matched = pi[matched - 1];
      }
    }
  }

  return ret;
}

int main(void) {
  default_random_engine gen(43);

  for (int it = 0; it < TEST_CASES; it++) {
    string h;
    string m;
    initialize(h, m, gen);

    vector<int> ret = solve(h, m);

    for (auto i : ret) {
      cout << i << " ";
    }
    cout << endl;
  }

  return 0;
}

File: python/algorithm/KMP_알고리즘.py
import random
from pprint import pprint

TEST_CASES = 10
H_SIZE = 1000000
M_SIZE = 10


def initialize():

    s = ["a", "b", "c"]

    h = "".join([random.choice(s) for _ in range(H_SIZE)])
    m = "".join([random.choice(s) for _ in range(M_SIZE)])

    return (h, m)


def get_partial_match(m):
    pi = [0 for _ in range(len(m))]

    for begin in range(1, len(m)):
        for i in range(len(m) - begin):
            if m[begin + i] != m[i]:
                break
            pi[begin + i] = max(pi[begin + i], i + 1)

    return pi


def solve(h, m):
    ret = []
    pi = get_partial_match(m)

    begin, matched = 0, 0

    while begin <= len(h) - len(m):
        if matched < len(m) and h[begin + matched] == m[matched]:
            matched += 1
            if matched == len(m):
                ret.append(begin)
        else:
            if matched == 0:
                begin += 1
            else:
                begin += matched - pi[matched - 1]
                matched = pi[matched - 1]

    return ret


if __name__ == "__main__":
    random.seed(43)

    for _ in range(TEST_CASES):
        h, m = initialize()
        ret = solve(h, m)
        print(ret)
